pkg/cache/priority_queue.go

```diff
@@ -0,0 +1,191 @@
+package cache
+
+import (
+	"container/heap"
+	"math"
+)
+
+const (
+	defaultCapacity = 8096
+)
+
+// PriorityQueue queue has priority, the first element has the highest priority
+type PriorityQueue struct {
+	queue    *priorityHeap
+	items    map[interface{}]*Entry
+	capacity int
+}
+
+// NewPriorityQueue construct of priority queue
+// -1: unlimited
+// 0: default capacity
+func NewPriorityQueue(capacity int) *PriorityQueue {
+	if capacity < 0 {
+		capacity = math.MaxInt16
+	} else if capacity == 0 {
+		capacity = defaultCapacity
+	}
+
+	return &PriorityQueue{
+		queue:    &priorityHeap{},
+		items:    make(map[interface{}]*Entry),
+		capacity: capacity,
+	}
+}
+
+// Push push element into pq with priority
+// it will refused if used > capacity
+func (pq *PriorityQueue) Push(priority int, x interface{}) bool {
+	v, ok := pq.items[x]
+	if !ok {
+		if pq.Size() >= pq.capacity {
+			return false
+		}
+		v = &Entry{Value: x, Priority: priority}
+		pq.queue.Push(v)
+	} else {
+		pq.Update(v, priority)
+	}
+	pq.items[x] = v
+	return true
+}
+
+// Pop pop the high Priority element from queue
+func (pq *PriorityQueue) Pop() *Entry {
+	entry, ok := pq.queue.Pop().(*Entry)
```

Comment:
```suggestion
	entry, ok := heap.Pop(pq.queue).(*Entry)
```
ditto

Reply:
it is different between heap. heap's element order by asc. I want to desc for get all.

Reply:
You defined `Less`, it is desc.